### PR TITLE
[sqlite] add a way to provide custom directory for the database

### PR DESCRIPTION
**Pull request by IgorKhramtsov**
# Why

Hey, thank you for the awesome ecosystem and packages! In the app I'm working on, I needed a way to share the SQLite database with a native widget, on iOS this could be done using shared containers, but that requires a way to provide a path for the SQLite database. 

This PR resolves the following feature request:
* https://expo.canny.io/feature-requests/p/support-for-setting-sqlite-database-path-to-ios-app-groups-directory

And partially resolves this feature request (a new function should be introduced in the `expo-file-system` package to get the path to the iCloud directory):
* https://expo.canny.io/feature-requests/p/sqlite-choose-location

# How

This PR does a few things in order to provide a way to store a database in an iOS shared container:
1. The new `directory` parameter is added, which will be combined with the `databaseName` to create a full path before supplying it to the native code. This way the backward compatibility is kept without introducing new methods;
2. The new `getSharedContainer` function is added in the `expo-file-system/next` package to get the path to a shared container;
3. Additional logic is added in the `getPathPermissions` function for paths in shared containers on iOS. Apparently, the permission check works differently in the shared container so when you try to check permissions of non-existent directory or parent directory - it will fail although there are enough permissions on the file system. The new logic will just check permissions in the root of the shared container;
4. New `ensureDatabasePathExists` (Sync and Async) functions are added in the SQLite native code to verify that the path exists and there are enough permissions before opening the database. Those functions basically are part of the constructor code moved out as during the tests I noticed that exceptions thrown in the constructor are not propagated correctly to the user (instead, it says that it was unable to find the native shared object associated with the given JS object; looks like it tries to map native Exception to the JS NativeDatabase or smth like that).

# Test Plan

New tests are added in the `test-suite` app for
1. `expo-file-system` to verify that it can create/delete files in the shared iOS container (and there are no permission issues);
2. `expo-sqlite` to verify that the database can be created opened and deleted in the custom directory and in the shared iOS container (I've added an app group in the bare-expo app for this test, but that might not work in the CI/CD if tests are executed from another app).

I, then, verified that those tests pass on iOS 17, 16, and Android 14.0 simulators using `bare-expo` app.
Also, I have tested this functionality in the app I'm working on by publishing forks of those two packages and using them in my project.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModuleNext.kt
Comment (IgorKhramtsov) [inline packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModuleNext.kt:210]: Previously the original `IOException` message was lost but it could be helpful for debugging/reporting

File: packages/expo-sqlite/ios/SQLiteModuleNext.swift
Comment (IgorKhramtsov) [inline packages/expo-sqlite/ios/SQLiteModuleNext.swift:51]: Here and below `absoluteString` is replaced with `standardizedFileUrl.path` because otherwise, it fails if the path points to the shared container

File: packages/expo-sqlite/package.json
Comment (Kudo) [inline packages/expo-sqlite/package.json:55]: having expo-file-system dependency may bring multiple expo-file-system versions in node_modules if people don't setup version correctly. i would recommend not introducing the other expo modules dependencies inside an expo module.

i looked into your pr and it turns out you were using expo-file-system only to get the document directory path. how about serving the feature from expo-sqlite directly? you can use [Constants](https://docs.expo.dev/modules/module-api/#constants)

**Review by Kudo — COMMENTED** [review]
thanks for having such fantastic pr. your work is great and so complete. before diving into detailed review, i have some first thought to share:
- introducing expo-file-system as a dependency to expo-sqlite is not ideal for our version management. please see the package.json inline comment.
- we are planning to have `expo-file-system/next` in the upcoming expo sdk release. how do you think if we add the `getSharedContainerUriAsync` to the `expo-file-system/next` and only available from the next version? cc @aleqsio

**Comment by IgorKhramtsov** [comment]
> we are planning to have expo-file-system/next in the upcoming expo sdk release. how do you think if we add the getSharedContainerUriAsync to the expo-file-system/next and only available from the next version? 

I'm working on it

**Comment by IgorKhramtsov** [comment]
@Kudo I've removed `getSharedContainerUriAsync` from `expo-file-system` and added `getSharedContainer` as a static function under `Directory` in the `expo-file-system/next`.

File: packages/expo-modules-core/ios/FileSystemUtilities/FileSystemUtilities.swift
Comment (Kudo) [inline packages/expo-modules-core/ios/FileSystemUtilities/FileSystemUtilities.swift:60]: i'm sorry for late to review the pr. i have little concern of having this ios internal path here. also that's not ideal to maintain appGroupId from both entitlment and javascript calls. i have some idea to try out. are you open to let me have a pr to work on App Group beforehand (will list you as co-author for sure).

File: packages/expo-sqlite/src/SQLiteDatabase.ts
Comment (Kudo) [inline packages/expo-sqlite/src/SQLiteDatabase.ts:446]: let me double check if that works on windows

File: packages/expo-sqlite/src/SQLiteDatabase.ts
Comment (Kudo) [inline packages/expo-sqlite/src/SQLiteDatabase.ts:462]: there's too many occurrences of  `createDatabasePath(databaseName, resolveDbDirectory(directory));`. maybe we can move this to`pathUtils.ts` and have `function createDatabasePath(databaseName: string, directory?: string` there. then we can also do some js unit test there

**Review by Kudo — COMMENTED** [review]
mostly looks great for expo-sqlite code changes. i will test windows support further and propose a pr for AppGroup change for expo-file-system

File: apps/test-suite/tests/FileSystemNext.ts
Comment (aleqsio) [inline apps/test-suite/tests/FileSystemNext.ts:290]: I'm working on an api for exposing common paths like `cache` and `documents` here:
https://github.com/expo/expo/pull/31467/

maybe this should be a property on `Paths`, like `Paths.getSharedContainer()`

File: packages/expo-file-system/ios/FileSystemExceptions.swift
Comment (aleqsio) [inline packages/expo-file-system/ios/FileSystemExceptions.swift:103]: `Check your entitlements`– could we make this more descriptive, like letting folks know which entitlements are missing or an FYI link?

**Review by aleqsio — APPROVED** [review]
Sorry for the late review, I like the FS part, but we'll need to update it to latest main and align with some recent changes. Happy to help out with this :)

**Comment by Kudo** [comment]
> i have some idea to try out. are you open to let me have a pr to work on App Group beforehand (will list you as co-author for sure).

having the pr at #31525, basically the appGroupId is passing from native (as single source of truth), so the callsite doesn't need to specify appGroupId.

```ts
  /**
   * Gets the shared container directories for [Apple App Groups](https://developer.apple.com/documentation/xcode/configuring-app-groups).
   * @return Map of app group identifiers to shared directories.
   * @platform ios
   * @platform tvos
   */
  static getAppleSharedContainersAsync(): Promise<Record<string, Directory>>;
```

**Comment by IgorKhramtsov** [comment]
> having the pr at https://github.com/expo/expo/pull/31525, basically the appGroupId is passing from native (as single source of truth), so the callsite doesn't need to specify appGroupId.

I checked your PRs, those changes look great! I will wait until they get merged to update my PR. As there are a few approvals on them already, I assume that will be quick.

**Comment by Kudo** [comment]
> I checked your PRs, those changes look great! I will wait until they get merged to update my PR. As there are a few approvals on them already, I assume that will be quick.

thanks for waiting. all prs are landed on main. the new file-system api is `Paths.appleSharedContainers`. please help to rebase and i'd love to review this pr.

File: apps/test-suite/tests/SQLite.ts
Comment (Kudo) [inline apps/test-suite/tests/SQLite.ts:3]:
```suggestion
import { Paths } from 'expo-file-system/next';
```

File: apps/test-suite/tests/SQLite.ts
Comment (Kudo) [inline apps/test-suite/tests/SQLite.ts:922]:
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const sharedContainerDir = sharedContainerRoot.uri + 'SQLite';
```

File: apps/test-suite/tests/SQLite.ts
Comment (Kudo) [inline apps/test-suite/tests/SQLite.ts:931]:
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const dbDirectory = sharedContainerRoot.uri + 'SQLite';
```

File: apps/test-suite/tests/SQLite.ts
Comment (Kudo) [inline apps/test-suite/tests/SQLite.ts:954]:
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const dbDirectory = sharedContainerRoot.uri + 'SQLite';
```

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (Kudo) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:386]:
```suggestion
import { Paths } from 'expo-file-system/next';
```

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (Kudo) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:383]:
```suggestion
Use [`appleSharedContainers`](filesystem-next/#applesharedcontainers) from the [`expo-file-system`](filesystem-next/) module to retrieve the path to the shared container:
```

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (Kudo) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:393]:
```suggestion
      return Object.values(Paths.appleSharedContainers)?.[0]?.uri;
```

**Review by Kudo — COMMENTED** [review]
revise for `Paths.appleSharedContainers` api

File: apps/test-suite/tests/FileSystemNext.ts
Comment (IgorKhramtsov) [reply apps/test-suite/tests/FileSystemNext.ts]: _all file-system changes are reverted_

File: packages/expo-file-system/ios/FileSystemExceptions.swift
Comment (IgorKhramtsov) [reply packages/expo-file-system/ios/FileSystemExceptions.swift]: _all file-system changes are reverted_

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:364]:
```suggestion
To share a database with other apps/extensions in the same App Group, you can use shared containers by following the steps below:
```

Update this to avoid using a fragmented second sentence so the readability is improved.

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:368]:
````suggestion
<Step label="1">

Configure the App Group in app config:

```json app.json
````

Nit: Using line breaks like the above here helps us to avoid any unknown markup rendering issues.
Let's change `app.json` to app config since this file can have different file formats in a project (example: app.config.js).

File: docs/pages/versions/unversioned/sdk/sqlite.mdx
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/sqlite.mdx:384]:
```suggestion

</Step>

<Step label="2">

Use [`Paths.appleSharedContainers`](filesystem-next/#applesharedcontainers) from the [`expo-file-system`](filesystem-next/) library to retrieve the path to the shared container:
```

We prefer using ["library" over module](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#glossary) as per our writing style guide

**Review by amandeepmittal — APPROVED** [review]
Thanks for incorporating docs changes 👍 

Please resolve the lint error in `sqlite.mdx` file.

File: packages/expo-sqlite/CHANGELOG.md
Comment (Kudo) [inline packages/expo-sqlite/CHANGELOG.md:12]:
```suggestion
```

**Review by Kudo — APPROVED** [review]
the changes looks all good to me. sorry again for the late review. please help to resolve conflicts and fix the lint error in sqlite.mdx. then we are good to merge. thanks for having the great change

**Comment by IgorKhramtsov** [comment]
@Kudo done

**Review by Kudo — APPROVED** [review]
thanks again for the fantastic work 👏

**Pull request merged**
